clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -3564,23 +3543,188 @@ static void handleSYCLIntelSchedulerTargetFmaxMhzAttr(Sema &S, Decl *D,
+// NotEqualToOne means all argument values are not one, and Unknown means that
+// the argument values are invalid.
+enum class ArgCheck { Unknown, EqualToOne, NotEqualToOne };
+static ArgCheck AreArgValuesToOne(const Expr *E, const Expr *E1,
+		                  const Expr *E2) {
+  // If any of the expression is value dependent, test is invalid.
+  const auto *XDimExpr = dyn_cast<ConstantExpr>(E);
+  const auto *YDimExpr = dyn_cast<ConstantExpr>(E1);
+  const auto *ZDimExpr = dyn_cast<ConstantExpr>(E2);
+  if (!XDimExpr || !YDimExpr || !ZDimExpr)
+    return ArgCheck::Unknown;
+
+  // Check the argument values.
+  return (XDimExpr->getResultAsAPSInt() == 1 ||
+          YDimExpr->getResultAsAPSInt() == 1 ||
+          ZDimExpr->getResultAsAPSInt() == 1)
+          ? ArgCheck::EqualToOne
+          : ArgCheck::NotEqualToOne;
+}
+
+void Sema::AddSYCLIntelMaxGlobalWorkDimAttr(Decl *D,
+                                            const AttributeCommonInfo &CI,
+                                            Expr *E) {
+  if (!E->isValueDependent()) {
+    // Validate that we have an integer constant expression and then store the
+    // converted constant expression into the semantic attribute so that we
+    // don't have to evaluate it again later.
+    llvm::APSInt ArgVal;
+    ExprResult Res = VerifyIntegerConstantExpression(E, &ArgVal);
+    if (Res.isInvalid())
+      return;
+    E = Res.get();
 
-  Expr *E = A.getArgAsExpr(0);
+    // This attribute requires a non-negative value.
+    if (ArgVal < 0) {
+      Diag(E->getExprLoc(), diag::err_attribute_requires_positive_integer)
+          << CI << /*non-negative*/ 1;
+      return;
+    }
+    // This attribute must be in the range [0, 3].
+    if (ArgVal > 3) {
+      Diag(E->getBeginLoc(), diag::err_attribute_argument_out_of_range)
+          << CI << 0 << 3 << E->getSourceRange();
+      return;
+    }
+    // Check to see if there's a duplicate attribute with different values
+    // already applied to the declaration.
+    if (const auto *DeclAttr = D->getAttr<SYCLIntelMaxGlobalWorkDimAttr>()) {
+      // If the other attribute argument is instantiation dependent, we won't
+      // have converted it to a constant expression yet and thus we test
+      // whether this is a null pointer.
+      if (const auto *DeclExpr = dyn_cast<ConstantExpr>(DeclAttr->getValue())) {
+        if (ArgVal != DeclExpr->getResultAsAPSInt()) {
+          Diag(CI.getLoc(), diag::warn_duplicate_attribute) << CI;
+          Diag(DeclAttr->getLoc(), diag::note_previous_attribute);
+        }
+        // Drop the duplicate attribute.
+        return;
+      }
+    }
+    // If the declaration has a SYCLIntelMaxWorkGroupSizeAttr or
+    // ReqdWorkGroupSizeAttr, check to see if they hold equal values
+    // (1, 1, 1) in case the value of SYCLIntelMaxGlobalWorkDimAttr
+    // equals to 0.
+    auto Check = [this, &CI](const auto *A) -> bool {
```

Comment:
Oh, this looks to be doing the same thing as `AreArgValuesToOne()`, so perhaps that should remain as a function and used here?